internal/operations/cache.go

```diff
@@ -0,0 +1,84 @@
+// Copyright © 2022 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operations
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/hyperledger/firefly/internal/log"
+	"github.com/hyperledger/firefly/pkg/fftypes"
+)
+
+type operationCacheKey struct{}
+type operationCache map[string]*fftypes.Operation
+
+func getOperationCache(ctx context.Context) operationCache {
+	ctxKey := operationCacheKey{}
+	cacheVal := ctx.Value(ctxKey)
+	if cacheVal != nil {
+		if cache, ok := cacheVal.(operationCache); ok {
+			return cache
+		}
+	}
+	return nil
+}
+
+func getCacheKey(op *fftypes.Operation) (string, error) {
+	opCopy := &fftypes.Operation{
+		Namespace:   op.Namespace,
+		Transaction: op.Transaction,
+		Type:        op.Type,
+		Plugin:      op.Plugin,
+		Input:       op.Input,
+	}
+	key, err := json.Marshal(opCopy)
```

Comment:
This feels like extra evidence to the point that we should treat the inputs as references, rather than full data wherever possible. As this is going to be a large string cache key that we are marshaling.

Reply:
... as an aside I still think serialized JSON would be a more efficient cache key than a SHA etc.

Reply:
Cool, that was my inkling as well but wanted to have someone else back me up. But agreed it's yet another reason to minimize the size of "Input" as much as is practical.